### ISSUE TITLE
some symbol couldn't found when "go to symbol in a file or workspace"

### ISSUE THREAD

**Issue by wgqimut** [microsoft/vscode-cpptools#4452]
**Type: LanguageService**

**Describe the bug**
- OS and Version: macOS version 10.15
- VS Code Version: Version: 1.39.2 
- C/C++ Extension Version:0.26.0 (Remote SSH CentOS Linux release 7.5.1804 (Core))

**To Reproduce**

when I go to symbol in a file(`X.c`) or workspace, some symbol can't found, even though "go to this symbol's definition" is ok, "find all this symbol's reference" is also working properly.

After setting cpptool's  log level to "debug",  and resetting intellisense database, I found tag parsing `X.c` file is okay, but tag parsing `X.h` encountered a error, maybe this reason make some symbol in X.c file couldn't found?

**Additional context**
Current Configuration:
```
{
    "name": "Linux",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [],
    "compilerPath": "/usr/local/bin/gcc",
    "cStandard": "c11",
    "cppStandard": "c++17",
    "intelliSenseMode": "gcc-x64",
    "compileCommands": "${workspaceFolder}/compile_commands.json",
    "compilerArgs": [],
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
```

**Comment by sean-mcmanus** [comment]
Can you provide more info on the symbols that can't be found? Is there a macro used that requires evaluation in order to parse? What is the error message?

Go to Def and Find All References may be using IntelliSense symbol info, which Go to Symbol doesn't use (i.e. it only does "tag parsing", which doesn't evaluate preprocessor macros).

**Comment by wgqimut** [comment]
@sean-mcmanus the symbol  is a function which uses some macro definition, I'm not sure which macro evaluation failed make the symbol founding failed, the whole tag parsing process's error message is just `Tag parsing XXX.c/h encountered a error, but it may not matter. Let us know if symbols in the file can't be found`, but for the file contains the missing symbol, the tag parsing is okay.

**Comment by sean-mcmanus** [comment]
You may need to add the "disruptive macro" to the a cpp.hint file: https://docs.microsoft.com/en-us/cpp/build/reference/hint-files?view=vs-2019 , i.e. if evaluating the macro is required in order to correctly parse the code, e.g.
```
#define START_NAMESPACE namespace MyProject {
#define END_NAMESPACE }
```
If you consistently get an error message when tag parsing the file with the definition in it, then our parser might be hitting an error. Are you able to provide a code sample that is causing our parser to fail to parse?

**Comment by wgqimut** [comment]
@sean-mcmanus  Finally, I figure out the failed reason, the parser can't parse `list_for_each_entry_safe` macro, the macro definition is in the `list.h`:
```c
#define list_for_each_entry_safe(pos, n, head, member)        \
  for (pos = list_entry((head)->next, typeof(*pos), member),  \
      n = list_entry(pos->member.next, typeof(*pos), member); \
       &pos->member != (head);                                \
       pos = n, n = list_entry(n->member.next, typeof(*n), member))
```
when I add  a cpp.hint file, and fill the `#define list_for_each_entry_safe` in this file, my missing symbol then can be found when I "go to symbol in a file".

**BTW**, when I add cpp.hint file and reset intellisense database, I have to restart vscode-server process, the reseting database then works, I think maybe this is a bug for cpp.hint.

the code sample is here:
```c
#include <stdio.h>
#include "list.h"

typedef struct {
  int i;
  struct list_head list;
} test_t;

int func() {
  test_t* t = NULL;
  test_t* tmp_t = NULL;
  struct list_head t_list;

  INIT_LIST_HEAD(&t_list);
  list_for_each_entry_safe(t, tmp_t, &t_list, list) { printf("test"); }
}
int main(void) {
  func();
}
```
the list.h can be found in [here](https://github.com/gluster/glusterfs/blob/master/libglusterfs/src/glusterfs/list.h)

the cpp.hint is:
```c
#define list_for_each_entry_safe
```

**Comment by sean-mcmanus** [comment]
My guess is the Reload Window is required because of caching of the cpp.hint (I see a 10 second cache, but maybe it could be cached for longer). Your or I could file a separate issue to investigate that if you want.

**Comment by wgqimut** [comment]
Why the tag parser can’t parse list_for_each_entry_safe? It can parse the other macro in the list.h, why not this one

**Comment by sean-mcmanus** [comment]
You mean why can't it parse it without the cpp.hint? It's because the tag parser doesn't not evaluate macros; therefore, any macros that require expansion in order to parse correctly will fail to parse without the cpp.hint.

**Comment by sean-mcmanus** [comment]
FYI, this is fixed with 1.10.3 (pre-release) -- no cpp.hint is required for this case.
